src/Controller/QrCodeController.php

```diff
@@ -0,0 +1,201 @@
+<?php
+
+/*
+ * UrlRewrite Bundle for Contao Open Source CMS.
+ *
+ * @copyright  Copyright (c) 2021, terminal42 gmbh
+ * @author     terminal42 <https://terminal42.ch>
+ * @license    MIT
+ */
+
+namespace Terminal42\UrlRewriteBundle\Controller;
+
+use Contao\Backend;
+use Contao\BackendTemplate;
+use Contao\CoreBundle\Exception\PageNotFoundException;
+use Contao\File;
+use Contao\Input;
+use Contao\StringUtil;
+use Contao\Widget;
+use Doctrine\DBAL\Connection;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Routing\Exception\InvalidParameterException;
+use Symfony\Component\Routing\Exception\MissingMandatoryParametersException;
+use Terminal42\UrlRewriteBundle\QrCodeGenerator;
+
+class QrCodeController
+{
+    /**
+     * @var Connection
+     */
+    private $connection;
+
+    /**
+     * @var QrCodeGenerator
+     */
+    private $qrCodeGenerator;
+
+    /**
+     * @var RequestStack
+     */
+    private $requestStack;
+
+    /**
+     * QrCodeController constructor.
+     */
+    public function __construct(Connection $connection, QrCodeGenerator $qrCodeGenerator, RequestStack $requestStack)
+    {
+        $this->connection = $connection;
+        $this->qrCodeGenerator = $qrCodeGenerator;
+        $this->requestStack = $requestStack;
+    }
+
+    /**
+     * Index view.
+     */
+    public function index(): Response
+    {
+        if (($request = $this->requestStack->getCurrentRequest()) === null
+            || !($id = $request->query->getInt('id'))
+            || ($rewriteData = $this->connection->fetchAssociative('SELECT * FROM tl_url_rewrite WHERE id=?', [$id])) === false
+            || !$this->qrCodeGenerator->validate($rewriteData)
+        ) {
+            throw new PageNotFoundException();
+        }
+
+        $routeParameters = [
+            'scheme' => $request->getScheme(),
+            'host' => null,
+        ];
+
+        $template = new BackendTemplate('be_url_rewrite_qr_code');
+        $template->backUrl = Backend::getReferer();
+
+        // Add form to the template
+        $this->addFormToTemplate($template, $request, $rewriteData, $routeParameters);
+
+        // Generate the QR code only if ALL parameters are set
+        if (!\in_array(null, $routeParameters, true)) {
+            $this->addQrCodeToTemplate($template, $rewriteData, $routeParameters);
+        }
+
+        return $template->getResponse();
+    }
+
+    /**
+     * Add QR code to the template.
+     */
+    private function addQrCodeToTemplate(BackendTemplate $template, array $rewriteData, array $routeParameters): void
+    {
+        try {
+            $data = $this->qrCodeGenerator->generate($rewriteData, $routeParameters);
+
+            // Create an image out of QR code so it can be easily downloaded
+            $file = new File(sprintf('assets/images/q/qr-%s.svg', md5($data['url'])));
```

Comment:
I don't really like tempering with the thumbnail folder of Contao. We should use the image services for that.
But tbh I think we don't even need to save it on disk. Can't we just have a controller that returns the SVG on the fly?

Reply:
I have improved it in 00a479b. Right now it works the way that the URL is generated when submitting a form in the backend, and passed on to the controller as base64 string to generate a QR code out of it. This actually means you can pass any content as base64 to be encoded, so I wonder if it's not a security problem.

On the other hand, otherwise I would have to pass the rewrite ID, host name and parameters. Then I would have to validate everything again and output the URL as QR code, which seems more secure though.

What do you think?

Reply:
There you go: 5cd057b